Some refactoring of the Strand enum adding encoding (string and char)…

vruano opened the pull request:
… functionality.

### Description

Some refactoring of the Strand enum adding encoding (string and char) functionality.

Strand is lacking encoding methods it only had a decoding one (toStrand(String)) and an attempt to use toString or name would return POSITIVE, NEGATIVE... when in practice one wants '+', '-'... 

It also deprecates 'toStrand(String)' for 'decode(String)' for name consistency sake (encode/decode) and 'Strand' in toStrand is kinda redundant... ideally one would like to name it valueOf but is impossible (nor desirable) to override valueOf(String). 

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

pshapiro4broad commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,102 @@
+     */
+    public static Strand decode(final char ch) {
+        switch (ch) {
+            case '+': return POSITIVE;
```
This duplicates the relationship already established in the enum definition. Instead you could write
```
for (Strand strand : values()) {
    if (strand.charEncoding == ch) {
        return strand;
    }
}
throw new ...
```

vruano replied:
yeah... I guess it could be kinda of faster with the switch and I am relaying on the unit-test to verify consistency at build-time. But at best is premature optimization so happy to revert to that.

pshapiro4broad replied:
If it turns out to be a performance issue, you can build a static `Map` using the enum values and also avoid the data duplication. Given that there are only three enum members it seems unlikely that this will be a bottleneck.

vruano replied:
In fact the map is likely to carry too much overhead.

pshapiro4broad replied:
I was curious so I wrote a short test. The test showed that switch is only slightly faster than the Map. Using a for loop is an order of magnitude slower than both the switch and map. So if you want to avoid duplication, Map seems to be the best solution.

https://gist.github.com/pshapiro4broad/1acabaa3fe74ac03007f15b989fe78a8

vruano replied:
Wow! Never thought of it... perhaps the culprit is values()... if one is allowed to change its contents then it makes sense that each invocation would instantiate a new array.

vruano replied:
If you are still curious and since you have the test code ready for action then you could try to statically cache the result of values().

pshapiro4broad replied:
The code is in the gist link in my comment, I'll update it when I have a chance, but you can update it as well.

I would expect that `Enum.values()` is a static reference since all enum instances are known at the time the .class file is loaded. It would be easy to try out.

lbergelson replied:
Arrays are always mutable, so I'm pretty sure it will allocate a fresh one on every call to values.

vruano replied:
I have done the test and that is the cause of the bad performance. Caching it into an static final bring time to the par with other solutions.

pshapiro4broad replied:
Good to know, thanks for checking this out.

pshapiro4broad commented on src/test/java/htsjdk/tribble/annotation/StrandTest.java:
```diff
@@ -0,0 +1,94 @@
+    public Object[][] invalidEncodingCharsData() {
+        final char[] testCases = new char[] {'c', '\t', '\n', (char) -1, 'p', 'n', 'x' };
+        return IntStream.range(0, testCases.length)
+                .mapToObj(i -> testCases[i])
```
Do you need `IntStream` here? It looks like you could use `Stream.of()` as do you for the string data.

vruano replied:
Yeah I guess i should just Use a ```Character[]``` array rather than a ```char``` array, perhaps implicitly boxed by Stream.of(Object ... o)

vruano commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,102 @@
-        encoding = str;
+    POSITIVE('+'),
+    /**
+     * Represent the negative or reverse strand.
```
Represents.

vruano commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,102 @@
+    NEGATIVE('-'),
+
+    /**
+     * Denotes that an strand designation is not applicable
```
an -> a.

magicDGS commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,102 @@
+     * Denotes that an strand designation is not applicable
+     * or is unknown.
+     */
+    NONE('!');  // not really sure what we should do for the NONE Enum
```
Now that this is refactored, maybe is a good time to add two instances for no strand to distinguish between non-available (not parsed, column not present in the file) and no strand (`.` in the BED format).

vruano replied:
I'm happy to change '!' for a '.' ... in what situation/format a '!' would mean no-strand? 

However I'm not sure about adding more NONEs... I would like to see a situation in where one would need to distinguish between more than one NONE.

In any case one could use ```Optional<Strand>``` to add an extra NONE... or even drop the current one!

yfarjoun commented on src/test/java/htsjdk/tribble/annotation/StrandTest.java:
```diff
@@ -0,0 +1,94 @@
+package htsjdk.tribble.annotation;
```
these tests aren't running for some reason...there is no tests coverage in Strand.java

vruano replied:
Just submitted a commit extending HtsjdkTest. Hopefully that fixes it.

vruano replied:
That did the trick, thanks!

yfarjoun commented on src/test/java/htsjdk/tribble/annotation/StrandTest.java:
```diff
@@ -0,0 +1,94 @@
+ *     Makes sure that {@link Strand} will behave according to contract in run-time.
+ * </p>
+ */
+public class StrandTest {
```
oh, you need to derive this class from HtsTestJava or something like that. take a look at one of the other files.

vruano commented:
Just submitted a commit addressing comments thus far. For decode I have opted for an updated for-each loop using a cached values arrays. 

For now I have not changed the NONE char representation from '!' to '.'. Please comment of whether you think that this change would be wise since it brakes back-comapatibily slightly. I have never seen '!' to be used to denote a No strand and '.' is commonly use to denote undefined values in common formats so it seems like a logical change.

Coverage test seems to do well now except for an unexpected change in coverage in an unchanged file... don't know why is that but it seems to be for the better so I guess it should not be a deterrent for a merge. 

There is a couple of uncovered lines in Strand (deprecated toStrand and toString) Will fix that before merging.

vruano commented:
Also I have added a couple alias/constants FORWARD = POSITIVE, REVERSE = NEGATIVE.

magicDGS commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,116 @@
+    }
+
+    @Override
+    public String toString() {
```
Overriding this method makes conversion of  `String` <-> `Strand` failed in an unexpected way (unless documented). For example, `Strand.valueOf(Strand.POSITIVE.toString())` won't return `Strand.POSITIVE` anymore. Because `valueOf` cannot be overriden, I think that the `toString()` method shouldn't return the encoding....

vruano replied:
mmm... looking at the standard Java doc for ```name()```  it seems as the overriding of toString() is actually encouraged to provide a more user friendly string representation of the enum constant. 

Also valueOf javadoc says explicitly is mean to recover constant by its constant name so to be used like so 
```
String x = X.name();
...
// millions lines later in another class far far a way:
...
EnumX X = EnumX.valueOf(y) // where somehow y came from x.
```
So assuming that ```valueOf(X.toString()) == X``` is a understandable misinterpretation  of what ```valueOf``` is supposed to do. I guess that that is why it makes sense for it to fail with an Exception rather than return a null as go are getting wrong the hard-wired programatic name for a constant.

That said, it is true that this change break back-copatability a bit and I don't feel very strong about it ...to show "strand = POSITIVE" rather than "strand = +" in error messages doesn't really make much of a difference.... an further enforces the use of ```encode``` and ```decode``` for reading it in or out of files... so I will revert that. 

Any other opinions.

magicDGS replied:
You're perfectly right, sorry. It is the value of `X.name()` the one that should be kept the same. I'm OK with the change, even if it is not backwards compatible -  in principle I am in favor of `toString()` methods only implemented for error/debugging messages and not rely on them for other purposes...

vruano commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,116 @@
+     * Denotes that a strand designation is not applicable
+     * or is unknown.
+     */
+    NONE('!');  // not really sure what we should do for the NONE Enum
```
I prefer the ```Optional<X>``` to handle that for those that really need to make such a distinction. The fact that there is as many non-proper enum constants (?, .) than proper one (+, -) is already smelly IMO. Anyone else can chime in? Can be de tie breaker?

magicDGS replied:
That's a change in the codecs, not in the enum,but I do not block the PR due to this discussion (I can propose a change in a following PR, either for the codec or for the enum). I am fine if only one "none" concept exists as far as it is encoded with the default for BED, because it is the codec which uses this currently (and also it is common in other formats).

For maintain backwards compatibility I guess that the best approach will be to deprecate `NONE` (to keep the `'!'` econding) in favor of a `MISSING`or `UNKNOWN` encoded with `'.'`. Once the `NONE` enum is removed, an alias can be added (similarly to the `POSITIVE`/`FORWARD` alias).

yfarjoun replied:
I would go with "."

yfarjoun replied:
where did the "!" suggestion come from?

magicDGS replied:
The `!` was the previous implementation... I guess that it comes from times when undefined wasn't `.` yet.

yfarjoun replied:
oh, OK. leave as is then!

magicDGS replied:
I would like to have `.`, because it is actually the representation for the BED file (at the moment, the only Strand usage within the library).

vruano replied:
Unless someone have show that '!' is actually a used in practice I'm inclined to change it to '.'

vruano replied:
I'm not going to add yet another "NONE" constant in this pull request.  @magicDGS if you fill strong about this I suggest that you work on another pull request with that change. Hopefully you can show an example how that that change improve some other piece of code in htsjdk.

magicDGS replied:
I was referring in the past comment to do exactly the change from '!' to '.', as you've already did. Thanks!

vruano commented on src/main/java/htsjdk/tribble/annotation/Strand.java:
```diff
@@ -24,29 +24,116 @@
+                return value;
+            }
+        }
+        throw new IllegalArgumentException("Unable to match encoding to Strand enum for encoding char: " + ch);
```
.... so that is even suggesting yet another constant INVALID? this is a dangerous slippery slope. I think this even strengths the point that no more non-proper enum constant should be added... tomorrow someone may thing of something else like "NOT_APPLICABLE"... "BOTH".... In this particular case I think is best to simply fail early if an invalid string is to be converted to a Strand

magicDGS replied:
I am fine with the exception too, it was just an idea...

yfarjoun commented on src/test/java/htsjdk/tribble/annotation/StrandTest.java:
```diff
@@ -0,0 +1,113 @@
+        for (final Strand value : Strand.values()) {
+            Assert.assertEquals(value.toString(), value.encode());
+        }
+    }
```
it would be god to have an explicit test that shows that you can decode valid chars and strings into the three difference `Strand`s.

yfarjoun commented:
an online search shows that BED format doesn't have a specified value for unspecified strand...however GFF (which in some-sense an extension of BED) does specify '.' as "unspecified".

I guess that's a 👍 from me.

please make sure that when *reading* a file you convert non-confoming values to "NONE" so that the odd BED file that has "!" will be read correctly.

vruano commented:
Now seems that is all in order. 

Now unrecognized encodings (including the '!') are decoded into NONE. 

Changes are mostly back-compatibly except for Strand#toString (uses the encoding rather than the constant name) and NONE canonic encoding is now '.' instead of '!'.

An unrelated issue with a URL change in ENA service was causing a test to fail, presumably master is also currently broken.

yfarjoun commented:
Looks good, though I'd ask that you remove the ENA modification and rebase once I merge that commit.

@lbergelson do you need to 👍 this?

vruano commented:
@lbergelson any updates?

lbergelson commented:
@vruano This looks good to me.  👍
